### PR TITLE
feat: more efficient data structure for checking Withdrawal Queue 

### PR DESCRIPTION
**Pull request by Niraj-Kamdar**
related: #152

### CODE REVIEW

File: contracts/Vault.vy [hunk 1]
```diff
@@ -553,36 +556,44 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+        assert key < SET_SIZE
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for m in range(key, key + SET_SIZE):
+            if m >= SET_SIZE:
+                for n in range(SET_SIZE):
+                    if n >= key:
+                        break
+                    if set[n] == queue[i]:
+                        assert False
```
Comment (Niraj-Kamdar) [inline contracts/Vault.vy:584]: duplicate found

Comment (fubuloubu) [reply contracts/Vault.vy:584]:
```suggestion
                        raise
```

**Comment by fubuloubu** [comment]
> ~~fixes~~ related: #152

I changed this to only being related, because we wanted to see if a number of other things might be possible e.g. unlimited size queue, less gas costs, etc.

**Comment by Niraj-Kamdar** [comment]
Can you tell me how to fix the commit lint issues?

File: contracts/Vault.vy [hunk 2]
```diff
@@ -553,36 +556,45 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
 
-        existsInOldQueue: bool = False
-        for j in range(MAXIMUM_STRATEGIES):
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # hash will have hash of the address in uint256.
+        # mask will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        mask: uint256 = SET_SIZE - 1
+        key: uint256 = bitwise_and(hash, mask)
```
Comment (fubuloubu) [inline contracts/Vault.vy:574]:
```suggestion
        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `SET_SIZE` bits of address
```

File: contracts/Vault.vy [hunk 3]
```diff
@@ -553,36 +556,45 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+        assert key < SET_SIZE
```
Comment (fubuloubu) [inline contracts/Vault.vy:575]: By definition this is correct
```suggestion
```

File: contracts/Vault.vy [hunk 4]
```diff
@@ -553,36 +556,45 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for m in range(key, key + SET_SIZE):
+            if m >= SET_SIZE:
+                for n in range(SET_SIZE):
+                    if n >= key:
+                        # Note: This only means SET_SIZE is lower than MAXIMUM_STRATEGIES
+                        raise
```
Comment (fubuloubu) [inline contracts/Vault.vy:583]: Prefer this
```suggestion
                    # Note: This only means SET_SIZE is lower than MAXIMUM_STRATEGIES
                    assert n < key
```

File: contracts/Vault.vy [hunk 5]
```diff
@@ -553,36 +556,45 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+                    if set[n] == queue[i]:
+                        raise
```
Comment (fubuloubu) [inline contracts/Vault.vy:585]:
```suggestion
                    assert set[n] != queue[i]
```

File: contracts/Vault.vy [hunk 6]
```diff
@@ -553,36 +556,45 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+                    if set[n] == ZERO_ADDRESS:
+                        set[n] = queue[i]
+                        break
+                break
+            if set[m] == queue[i]: 
+                raise
```
Comment (fubuloubu) [inline contracts/Vault.vy:591]:
```suggestion
            assert set[m] == queue[i]
```

**Review by fubuloubu — COMMENTED** [review]
What happens when there is a hash collision in the set because only 5 bits of entropy (`2**5 = 32`) are used?

**Comment by Niraj-Kamdar** [comment]
> What happens when there is a hash collision in the set because only 5 bits of entropy (`2**5 = 32`) are used?

Since MAXIMUM_STRATEGIES is 20,  SET_SIZE used is 32 (nearest bigger power of 2)
Whenever hash collision happen. We will find the next empty slot in the SET which will take linear time and since SET_SIZE will always be greater than MAXIMUM_STRATEGIES, it will be able to find an empty slot for sure. Increasing the SET_SIZE can increase the performance if there are many collision happening but probability of multiple collision reduces very quickly. So, I don't think it will be an issue. 

If we want to be extra cautious then we can hash all the possible addresses and find out exactly how many collisions can happen.

**Comment by fubuloubu** [comment]
> > What happens when there is a hash collision in the set because only 5 bits of entropy (`2**5 = 32`) are used?
> 
> Since MAXIMUM_STRATEGIES is 20, SET_SIZE used is 32 (nearest bigger power of 2)
> Whenever hash collision happen. We will find the next empty slot in the SET which will take linear time and since SET_SIZE will always be greater than MAXIMUM_STRATEGIES, it will be able to find an empty slot for sure. Increasing the SET_SIZE can increase the performance if there are many collision happening but probability of multiple collision reduces very quickly. So, I don't think it will be an issue.
> 
> If we want to be extra cautious then we can hash all the possible addresses and find out exactly how many collisions can happen.

Ah, I think I get it now. Basically, it just orders the set by the first 5 bits of the hash.

Honestly, the address is already a hash, could probably just take the first 5 bits of that

**Comment by Niraj-Kamdar** [comment]
> Honestly, the address is already a hash, could probably just take the first 5 bits of that
Don't we need to convert it to uint?
If it would be possible to get the last 5 bits from the address directly. It will be super efficient.

**Comment by fubuloubu** [comment]
> Can you tell me how to fix the commit lint issues?

`git rebase -i master` and then edit or squash commits as needed

**Comment by fubuloubu** [comment]
> > Honestly, the address is already a hash, could probably just take the first 5 bits of that
> > Don't we need to convert it to uint?
> > If it would be possible to get the last 5 bits from the address directly. It will be super efficient.

Just bitshift by 155 bits (address is bytes20 aka 160 bits, but you only need 5 bits for the ordered set)

**Comment by Niraj-Kamdar** [comment]
> Just bitshift by 155 bits (address is bytes20 aka 160 bits, but you only need 5 bits for the ordered set)
I will try this out and compare the gas profile.

**Comment by fubuloubu** [comment]
> > Just bitshift by 155 bits (address is bytes20 aka 160 bits, but you only need 5 bits for the ordered set)
> 
> I will try this out and compare the gas profile.

Oh wait, you are already doing this lol. I thought you added an extra hash in there, which is like 700 gas

**Comment by Niraj-Kamdar** [comment]
> Oh wait, you are already doing this lol. I thought you added an extra hash in there, which is like 700 gas
Nope just type casting and taking bitwise_and

File: contracts/Vault.vy [hunk 7]
```diff
@@ -553,36 +556,40 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+        # `mask = SET_SIZE - 1` will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `SET_SIZE` bits of address
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for m in range(key, key + SET_SIZE):
+            if m >= SET_SIZE:
+                for n in range(SET_SIZE):
+                    # Note: This only means SET_SIZE is lower than MAXIMUM_STRATEGIES
+                    assert n < key
```
Comment (fubuloubu) [inline contracts/Vault.vy:580]: Would this be appropriate, considering the above comment?
```suggestion
                    assert n <= MAXIMUM_STRATEGIES
```

Comment (Niraj-Kamdar) [reply contracts/Vault.vy:580]: Nope it should be `<key` because `>= key` has been traversed already. Comment is for understanding the situation better.

File: contracts/Vault.vy [hunk 8]
```diff
@@ -553,36 +556,40 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # hash will have hash of the address in uint256.
+        # `mask = SET_SIZE - 1` will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `SET_SIZE` bits of address
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for m in range(key, key + SET_SIZE):
+            if m >= SET_SIZE:
```
Comment (fubuloubu) [inline contracts/Vault.vy:577]: Add a note about why this works

File: contracts/Vault.vy [hunk 9]
```diff
@@ -553,36 +556,40 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+                for n in range(SET_SIZE):
+                    # Note: This only means SET_SIZE is lower than MAXIMUM_STRATEGIES
+                    assert n < key
+                    assert set[n] != queue[i]
+                    if set[n] == ZERO_ADDRESS:
+                        set[n] = queue[i]
+                        break
+                break
```
Comment (fubuloubu) [inline contracts/Vault.vy:585]: Doesn't this mean it exhausted the set available space without finding a place for the entry?

**Review by fubuloubu — COMMENTED** [review]
Overall, I still think there is a potential vulnerability if you attempt to add 2 or more strategies that compute the same key.

Could you please add a test case showing this is not the case? You might have to farm the strategy addresses by nonce farming (keep re-deploying the strategies until you get an address that satisfies this condition), or you can try adding a fuzzer case that tests this logic (e.g. `addStrategy`, `migrateStrategy` and `removeStrategy` as well as all the queue manipulation functions as rules)

File: contracts/Vault.vy (hunk 9 above)
Comment (Niraj-Kamdar) [reply contracts/Vault.vy:585]: Nope. it will only get triggered when it founds an empty space within the set. other cases will raise an error due to assertions.

**Comment by Niraj-Kamdar** [comment]
SET is designed to handle collisions efficiently as long as `SET_SIZE >= MAXIMUM_STRATEGIES` meets and `SET_SIZE` is power of 2. 

** Why  `SET_SIZE >= MAXIMUM_STRATEGIES`?**
You will need enough space in SET to incorporate maximum number strategies in the worst case scenario.

** Why power of 2 in particular?**
It will be computationally efficient and more collision resistant. 
Ex: `SET_SIZE = 32` => `mask = 31 = b"11111"`, taking `bitwise_and` with the `int` representation of address will give us key. So SET_SIZE with power of 2 will utilise all the bits efficiently. 

If we take `SET_SIZE = 20` then `mask=b"10011"`  => In this case, if `address1 = b"11100"` and `address2=b"10000"` will produce same key and create collision even with 2 different bits but having SET_SIZE as power of 2 length will solve this issue.

** How are you handling collision within the set?
Whenever collision happens, set will search for the next empty slot from the key position linearly. If it won't find it till the end which is the SET_SIZE then it will loop back to the beginning of set and try to find the empty key slot from the 0th index to key position. It will surely find one if `SET_SIZE >= MAXIMUM_STRATEGIES`.

**Comment by Niraj-Kamdar** [comment]
I can add functional test case showing that collision won't be a problem.
Testing collision: 
create 2 addresses with same 5 least significant bits and random 155 most significant bits and convert it to the address.

**Comment by Niraj-Kamdar** [comment]
I will use https://www.ethereumaddressgenerator.com/ for generating addresses with common suffixes.

File: contracts/Vault.vy [hunk 10]
```diff
@@ -553,36 +556,40 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
         assert self.strategies[queue[i]].activation > 0
 
-        existsInOldQueue: bool = False
-        for j in range(MAXIMUM_STRATEGIES):
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # hash will have hash of the address in uint256.
+        # `mask = SET_SIZE - 1` will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `SET_SIZE` bits of address
```
Comment (fubuloubu) [inline contracts/Vault.vy:573]:
```suggestion
        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `log_2(SET_SIZE)` bits of address
```

File: contracts/Vault.vy (hunk 8 above)
Comment (fubuloubu) [reply contracts/Vault.vy:577]:
```suggestion
            # We've exhausted the set but still need to place this item
            if m >= SET_SIZE:
```

File: contracts/Vault.vy [hunk 11]
```diff
@@ -553,36 +556,40 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for m in range(key, key + SET_SIZE):
+            if m >= SET_SIZE:
+                for n in range(SET_SIZE):
```
Comment (fubuloubu) [inline contracts/Vault.vy:578]:
```suggestion
                # So we're starting over from the beginning of the set to find space
                for n in range(SET_SIZE):
```

File: contracts/Vault.vy (hunk 9 above)
Comment (fubuloubu) [reply contracts/Vault.vy:585]: This is not strictly true actually, the `break` internal to the `if` statement above it should trigger if it finds space in the set, this line needs to change to be a `raise` to account for the scenario where no space is found at all
```suggestion
                # NOTE: should not revert if SET_SIZE is lower than MAXIMUM_STRATEGIES
                raise UNREACHABLE  # dev: no space in set found
```

File: contracts/Vault.vy (hunk 7 above)
Comment (fubuloubu) [inline contracts/Vault.vy:580]: Delete this line to save a tiny amount of gas now that we raise if the set is filled up prematurely
```suggestion
```

Comment (fubuloubu) [reply contracts/Vault.vy:580]: Yes, this will waste a little extra gas getting to the final conclusion, but this invariant will still be maintained by proxy, and also it reduces costs in the common case workflow of 1-2 passes before filling in the set

**Review by fubuloubu — APPROVED** [review]
Some more improvements and documentation needed, then LGTM

**Comment by fubuloubu** [comment]
> I will use https://www.ethereumaddressgenerator.com/ for generating addresses with common suffixes.

Might be more difficult, the address has to be a valid strategy (e.g. has `.vault()` and `.want()`) to be added to the queue. There might be a way to mock this, but it's probably not trivial. I'd recommend the fuzzing case anyways.

**Comment by Niraj-Kamdar** [comment]
No outer `break` shouldn't be replaced with the raise and `assert n < key` is very necessary.

**Comment by fubuloubu** [comment]
> No outer `break` shouldn't be replaced with the raise and `assert n < key` is very necessary.

Please explain, I don't see it

**Comment by Niraj-Kamdar** [comment]
Ex: `key = 30` and both 30 and 31 is filled. Now we start with 0th slot inside the `if m >= SET_SIZE` part, and start looping from start to find empty space if we find the same element at any place we flag it as duplicate and raise an error. if we find an empty slot we put the strategy address there and break the **inner loop**. control goes to the if block outside the inner loop which is inside another loop which you need to break and register the strategy in withdrawal queue.

If you raise an error there a valid transaction would revert.

**Comment by Niraj-Kamdar** [comment]
> Might be more difficult, the address has to be a valid strategy (e.g. has `.vault()` and `.want()`) to be added to the queue. There might be a way to mock this, but it's probably not trivial. I'd recommend the fuzzing case anyways.

I don't know how to do it.  can you help me?

File: contracts/Vault.vy (hunk 9 above)
Comment (fubuloubu) [reply contracts/Vault.vy:587]: > Ex: `key = 30` and both 30 and 31 is filled. Now we start with 0th slot inside the `if m >= SET_SIZE` part, and start looping from start to find empty space if we find the same element at any place we flag it as duplicate and raise an error. if we find an empty slot we put the strategy address there and break the **inner loop**. control goes to the `if block` outside the inner loop which is inside another loop which you need to break and register the strategy in withdrawal queue.
> 
> If you raise an error there a valid transaction would revert.

Ah, I see the difference now.

The change would be this:
```suggestion
                    # NOTE: should not revert if SET_SIZE is lower than MAXIMUM_STRATEGIES
                    raise UNREACHABLE  # dev: no space in set found

                # NOTE: break out of top loop
                break
```
With this change, removing `assert n < key` above would be valid as that invariant check happens here.

---

Alternatively, one could just `m %= SET_SIZE` on the first line to re-align the set index to be within the loop's boundaries

**Review by fubuloubu — COMMENTED** [review]
Responded inside the relevant conversation

**Comment by fubuloubu** [comment]
> > Might be more difficult, the address has to be a valid strategy (e.g. has `.vault()` and `.want()`) to be added to the queue. There might be a way to mock this, but it's probably not trivial. I'd recommend the fuzzing case anyways.
> 
> I don't know how to do it. can you help me?

The fuzzing case would be similar to others in the `tests/integration/` section. This section of Brownie's docs will help: https://eth-brownie.readthedocs.io/en/stable/tests-hypothesis-stateful.html

---

Alternatively, you can just keep deploying until you get two strategies with a key of 31. That would test the boundary condition. Add all the intermediate strategies in there (up to 20) for funsies.

**Comment by Niraj-Kamdar** [comment]
No,
```python
                break               
 # NOTE: should not revert if SET_SIZE is lower than MAXIMUM_STRATEGIES                
raise UNREACHABLE  # dev: no space in set found
```
Above would be wrong as well.

Idea of using modulo seems interesting though.

**Comment by fubuloubu** [comment]
> No, replacing `break` with this...
> 
> ```python
>  # NOTE: should not revert if SET_SIZE is lower than MAXIMUM_STRATEGIES                
> raise UNREACHABLE  # dev: no space in set found
> ```
> 
> would be wrong as well.

It would work in context. It should *always* break out of the inner loop, unless the set is full (which it shouldn't be based on the parameterization, but... the `raise` catches that), and then the next break statement throws you out of the outer loop as well, once you break the inner loop.

**Comment by fubuloubu** [comment]
> Idea of using modulo seems interesting though.

The modulo gets you over having the do the inner and outer loop setup

**Comment by Niraj-Kamdar** [comment]
> The modulo gets you over having the do the inner and outer loop setup
Yes but it seems using modulo is more gas expensive because of memory allocation and computation then just using `if` statement.

**Comment by Niraj-Kamdar** [comment]
> It would work in context. It should _always_ break out of the inner loop, unless the set is full (which it shouldn't be based on the parameterization, but... the `raise` catches that), and then the next break statement throws you out of the outer loop as well, once you break the inner loop.

assert at the beginning of inner loop would catch that and if we keep `raise ` in the end it will stop the loop in one round if it can't find an element. We wouldn't want to stop it until we exhaust space, find duplicate or an empty slot.

**Comment by fubuloubu** [comment]
> > It would work in context. It should _always_ break out of the inner loop, unless the set is full (which it shouldn't be based on the parameterization, but... the `raise` catches that), and then the next break statement throws you out of the outer loop as well, once you break the inner loop.
> 
> assert at the beginning of inner loop would catch that and if we keep `raise ` in the end it will stop the loop in one round if it can't find an element. We wouldn't want to stop it until we exhaust space, find duplicate or an empty slot.

This isn't like Python, it'll stop the whole transaction from progressing. It stopping from the `assert` or from the `raise` doesn't change the outcome, only the amount of gas it uses. Using the `assert` means it will use that much gas on every iteration, changing to the `raise` will only use that much gas in the case where the set is full, which a) should definitely not be the common case, and b) should never happen anyways, due to the parameterization. Plus, it makes the code look a lot cleaner IMO.

**Comment by fubuloubu** [comment]
> > The modulo gets you over having the do the inner and outer loop setup
> 
> Yes but it seems using modulo is more gas expensive because of memory allocation and computation then just using `if` statement.

Do you have numbers that show this? Would think it would be more expensive to do the if statements instead of just doing `m %= SET_SIZE`. Surely the codesize would be smaller from removing the 2nd loop.

**Comment by Niraj-Kamdar** [comment]
> Do you have numbers that show this? Would think it would be more expensive to do the if statements instead of just doing `m %= SET_SIZE`. Surely the codesize would be smaller from removing the 2nd loop.

Well `m` will be immutable so we need to declare new variable and assign it `m % SET_SIZE`. i think that's why it's expensive?

**Comment by fubuloubu** [comment]
> > Do you have numbers that show this? Would think it would be more expensive to do the if statements instead of just doing `m %= SET_SIZE`. Surely the codesize would be smaller from removing the 2nd loop.
> 
> Well `m` will be immutable so we need to declare new variable and assign it `m % SET_SIZE`. i think that's why it's expensive?
> 
> These are the numbers with `if`:
> 
> ```
>    ├─ setWithdrawalQueue              -  avg:   66567  avg (confirmed):   93818  low:   28438  high:  179829
> ```
> 
> and these with `%`:
> 
> ```
>    ├─ setWithdrawalQueue              -  avg:   66597  avg (confirmed):   93855  low:   28438  high:  179921
> ```
> 
> Not that much but still depending on gas price it can become expensive

Yeah, that's basically nothing lol. Can you post a diff with this change in play? I think the increased readability is worth the extra 43+ gas spent.

**Comment by Niraj-Kamdar** [comment]
> Yeah, that's basically nothing lol. Can you post a diff with this change in play? I think the increased readability is worth the extra 43+ gas spent.
Yeah it looks much nicer

File: contracts/Vault.vy [hunk 12]
```diff
@@ -553,36 +556,32 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
-                existsInOldQueue = True
+        # hash will have hash of the address in uint256.
+        # `mask = SET_SIZE - 1` will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `log_2(SET_SIZE)` bits of address
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for j in range(key, key + SET_SIZE):
+            idx: uint256 = j % SET_SIZE  # traverse set for empty space in circular manner
+            assert set[idx] != queue[i]  # duplicate check
```
Comment (fubuloubu) [inline contracts/Vault.vy:578]: This notation is specific to brownie
```suggestion
            assert set[idx] != queue[i]  # dev: duplicate in set
```

**Review by fubuloubu — APPROVED** [review]
This looks so much better!

File: contracts/Vault.vy (hunk 12 above)
Comment (fubuloubu) [reply contracts/Vault.vy:578]: With this notation, you can do `with brownie.reverts("dev: duplicate in set"): vault.addStrategy(...)` to make sure it fails on this specific line when adding a duplicated strategy

File: contracts/Vault.vy [hunk 13]
```diff
@@ -553,36 +556,32 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
-        for j in range(MAXIMUM_STRATEGIES):
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # hash will have hash of the address in uint256.
+        # `mask = SET_SIZE - 1` will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `log_2(SET_SIZE)` bits of address
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for j in range(key, key + SET_SIZE):
```
Comment (fubuloubu) [inline contracts/Vault.vy:576]:
```suggestion
        # NOTE: `key` is first `log_2(SET_SIZE)` bits of address (which is a hash)
        hash: uint256 = convert(queue[i], uint256)
        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  
        # Most of the times following for loop only run once which is making it highly gas efficient 
        # but in the worst case of key collision it will run linearly and find first empty slot in the set.
        for j in range(key, key + SET_SIZE):
```

File: contracts/Vault.vy [hunk 14]
```diff
@@ -553,36 +556,32 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
+            idx: uint256 = j % SET_SIZE  # traverse set for empty space in circular manner
```
Comment (fubuloubu) [inline contracts/Vault.vy:577]:
```suggestion
            # NOTE: we can always find space by treating set as circular (as long as `SET_SIZE >= MAXIMUM_STRATEGIES`)
            idx: uint256 = j % SET_SIZE
```

**Review by fubuloubu — COMMENTED** [review]
Clean up docs a bit more

File: contracts/Vault.vy [hunk 15]
```diff
@@ -553,36 +556,32 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
 
         assert self.strategies[queue[i]].activation > 0
 
-        existsInOldQueue: bool = False
-        for j in range(MAXIMUM_STRATEGIES):
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # NOTE: `key` is first `log_2(SET_SIZE)` bits of address (which is a hash)
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)
```
Comment (fubuloubu) [inline contracts/Vault.vy:572]: Don't really need the intermediate step here, save some gas
```suggestion
        key: uint256 = bitwise_and(convert(queue[i], uint256), SET_SIZE - 1)  
```

File: contracts/Vault.vy [hunk 16]
```diff
@@ -553,36 +556,32 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
...
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # NOTE: `key` is first `log_2(SET_SIZE)` bits of address (which is a hash)
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot in the set.
+        for j in range(key, key + SET_SIZE):
+            # NOTE: we can always find space by treating set as circular (as long as `SET_SIZE >= MAXIMUM_STRATEGIES`)
+            idx: uint256 = j % SET_SIZE
```
Comment (fubuloubu) [inline contracts/Vault.vy:577]: Curious if this might be a little nicer
```suggestion
        for j in range(SET_SIZE):
            # NOTE: we can always find space by treating set as circular (as long as `SET_SIZE >= MAXIMUM_STRATEGIES`)
            idx: uint256 = (key + j) % SET_SIZE
```

**Review by fubuloubu — COMMENTED** [review]
Sorry, keep having optimization ideas here

**Review by fubuloubu — APPROVED** [review]
LGTM

**Comment by fubuloubu** [comment]
Note: because of the reduction in complexity, this change seemed like value-add.

Still should add a fuzzing test case for queue management that ensures there's no way to violate the invariants in `setWithdrawalQueue`
